Descriptive RPC Interface Spec for Payment Channel Endpoint

manoranjith opened the pull request:
I have added a descriptive version of the API spec for rpc interface. Once this is finalized, we can derive protocol specific spec such as protocol buffers for grpc or json scheme for json rpc.

Relates to direct-state-transfer/dst-go#45

Signed-off-by: Manoranjith <ponraj.manoranjitha@in.bosch.com>

danielksan81 commented on design/001-RPC-Interface-Specification.md:
```diff
@@ -0,0 +1,413 @@
+
+* `Node config`: (list of configurable parameters is yet to be defined)
+
+#### 2. Open Session
+
+Open a new session for the given user with the specified configuration file.
+
+*Parameters*
+
+* `Config File`: [String] Path to the config file for the session. This should
+   present on a file system path accesible by the node.
```
Does it make sense also to pass the needed values instead of passing a file? As an alternative

manoranjith replied:
The following are the required parameters: on-chain credentials, off-chain credentials and database directory (for persisting data across sessions). Credentials include: address, path to keystore file and password.
and the following are optional parameters: chain URL, addresses of on-chain state channel contracts.

Since the password is included in the credentials (a required parameter), i think it is safe to specify it in a file and pass the path to it.

Maybe the addresses and optional parameters can be specified by the user over the API. However, since a file is required, I thought all the configuration can be provided via the file. 

Do you see benefits for passing part of the configuration via API ?

danielksan81 replied:
You're right, I forgot the password. Then it will make only sense later if the password/signing is not handled by the node but an external secured service.

manoranjith replied:
Yes, When we add the support for making signatures via external service, then we could add an option to provide all values via API itself.

danielksan81 commented on design/001-RPC-Interface-Specification.md:
```diff
@@ -0,0 +1,413 @@
+
+*Return*
+
+* `Open channels`: [List of channel]
```
List of channel ids or objects? I expect to have channel ids

manoranjith replied:
It is list of `channel`. The channel structure is defined in line 84. It includes:
1. Channel ID
2. Channel Balance and
3. Current Version

danielksan81 commented on design/001-RPC-Interface-Specification.md:
```diff
@@ -0,0 +1,413 @@
+1. Node APIs - For accessing node related functionality of the node. Following
+APIs should be provided:
+
+    1. Get config
+    2. Open Session
+    3. Help
```
Does it make sense to provide a list of available remote interface implementations / commands available and they will provide a self description if possible?
Like if they will run on the same port but different paths or maybe different ports...

manoranjith replied:
Yes, I think it makes sense to provide a list of APIs provided by a node instance. Currently I have updated the spec to include this information in the `Help` API.

danielksan81 commented on design/001-RPC-Interface-Specification.md:
```diff
@@ -0,0 +1,413 @@
+
+* `Unknown Session ID`
+
+#### 5. Subscribe To Payment Channel Requests
+
+Subscribe to notifications on new incoming payment channel requests.  A
+subscription id will be returned immediately and it will be included in the
+notification sent to the user on incoming payment channel requests. 
```
How is this working and where is the channel request coming to?
Or is this more or less "open" to the implementation, so that gRPC for example needs to specify a dedicated method/callback, which will be the same for a webhook implementation?
Same for the other subscription in a channel below.

manoranjith replied:
The specification itself tells that, 
1. Client should be able to register for specific updates. Eg. Incoming channel updates. and
2. Server should be able to send these updates to the client.

As you said, how these is done, is open to the implementation. Some possible ways are
* server streaming in gRPC
* web hooks
* persistent, bidirectional connection using websockets etc.,

manoranjith commented:
Based on the learning from implementation design, I have update the specification. The following are the significant changes.
#### Relating to Subscriptions:
1. Add a `timeout` field in the notifications on channel proposal requests and payment requests. The user should respond before this timeout expires.
2. Added a `Time` API that returns of the time as per the clock of perun node. This time is required to check if the `timeout` mentioned in the notification has expired.
3. Added APIs for unsubscribing an active subscription.
4. Added APIs to subscribe/unsubscribe for and receive notifications for channel close events.
5. Added more description to explain how subscriptions work.

#### Other changes:
5. Added  `Close Session` API for closing a session .
6. Removed `Get Contacts` (returns all contacts) API and added `Get Contact` API that returns the contact details for a peer corresponding to a given alias. This is because the purpose of contacts is to facilitate the user to refer to a peer via `Alias` and hence the user might need only contacts corresponding to an alias known to the user.
7. Added parameters for `Get Config` and `Help` APIs.
8. Added more errors with specific meanings.

manoranjith commented:
Updated few terminologies and added version to response of two APIs based on offline review with @anagha-ks.
See commit 8b78fa2.

1. Use Payee to refer the person being paid. Only positive values are
allowed for amount.
2. Use "Payment Channel Proposal" instead of "Payment Channel Request".
3. Use "Payment Channel Update" instead of "Payment".
4. Add version to response of "Get Balance" and "Close Payment Channel".
5. Add optional error field to response of "Close Payment Channel".
6. Change String to List of String in Help response.

anagha-ks commented:
Please squash all commits.
